src/picture.rs

```diff
@@ -148,8 +151,72 @@ impl Default for Rav1dPicAllocator {
 
 unsafe extern "C" fn free_buffer(_data: *const u8, user_data: *mut c_void) {
     let pic_ctx: *mut pic_ctx_context = user_data as *mut pic_ctx_context;
-    (*pic_ctx).allocator.release_picture(&mut (*pic_ctx).pic);
-    free(pic_ctx as *mut c_void);
+    let pic_ctx = Box::from_raw(pic_ctx);
+    let data = &pic_ctx.pic.data;
+    pic_ctx
+        .allocator
+        .dealloc_picture_data(data.data, data.allocator_data);
+}
+
+impl Rav1dPicAllocator {
+    pub fn alloc_picture_data(
+        &self,
+        w: c_int,
+        h: c_int,
+        seq_hdr: Arc<DRav1d<Rav1dSequenceHeader, Dav1dSequenceHeader>>,
+        frame_hdr: Option<Arc<DRav1d<Rav1dFrameHeader, Dav1dFrameHeader>>>,
+    ) -> Rav1dResult<Rav1dPicture> {
+        let pic = Rav1dPicture {
+            p: Rav1dPictureParameters {
+                w,
+                h,
+                layout: seq_hdr.layout,
+                bpc: 8 + 2 * seq_hdr.hbd,
+            },
+            seq_hdr: Some(seq_hdr),
+            frame_hdr,
+            ..Default::default()
+        };
+        let mut pic_c = pic.to::<Dav1dPicture>();
+        // Safety: `pic_c` is a valid `Dav1dPicture` with `data`, `stride`, `allocator_data` unset.
+        let result = unsafe { (self.alloc_picture_callback)(&mut pic_c, self.cookie) };
+        result.try_to::<Rav1dResult>().unwrap()?;
+        let mut pic = pic_c.to::<Rav1dPicture>();
+
+        let pic_ctx = Box::new(pic_ctx_context {
+            allocator: self.clone(),
+            pic: pic.clone(),
+        });
+        // Safety: TODO(kkysen) Will be replaced by an `Arc` shortly.
+        pic.r#ref = NonNull::new(unsafe {
+            rav1d_ref_wrap(
+                pic.data.data[0] as *const u8,
+                Some(free_buffer),
+                Box::into_raw(pic_ctx).cast(),
+            )
+        });
+        assert!(pic.r#ref.is_some()); // TODO(kkysen) Will be removed soon anyways.
+
+        Ok(pic)
+    }
+
+    pub fn dealloc_picture_data(
+        &self,
+        data: [*mut c_void; 3],
+        allocator_data: Option<NonNull<c_void>>,
+    ) {
+        let data = data.map(NonNull::new);
+        let mut pic_c = Dav1dPicture {
+            data,
+            allocator_data,
+            ..Default::default()
+        };
```

Comment:
It does slightly change the API, given those other fields are accessible, though I don't think they're supposed to be read.  From all of the implementations I looked at, none of them accessed other fields, though.  Including the other fields here is hard because then we can't `impl Drop` without moving those other fields inside `Rav1dPictureData`, which we can't do since then they'd have to be inside the `Arc`, which they're not supposed to be.

Reply:
I added this clarification of the APIs a66caa9.  Does that sound good enough?  I don't know how to make the API safe and keep exactly the existing unspecified but possible API.  I was hoping to get this merged soon so it doesn't get bogged down with a lot of merge conflicts later and it was basically already finished.